snap/local/runtime-helpers/bin/kong-postgres-setup.sh

```diff
@@ -0,0 +1,207 @@
+
+  iter_num=0
+  MAX_POSTGRES_INIT_ITERATIONS=10
+  until "$SNAP/bin/drop-snap-daemon.sh" "$SNAP/bin/perl5lib-launch.sh" "$SNAP/usr/bin/createdb" kong; do
+      sleep 1
+      iter_num=$(( iter_num + 1 ))
+      if [ $iter_num -gt $MAX_POSTGRES_INIT_ITERATIONS ]; then
+          logger "edgexfoundry:postgres: failed to create kong db in postgres after $iter_num iterations"
+          exit 1
+      fi
+  done
+}
+
+postgres_create_directories() {
+
+     logger "edgexfoundry:postgres: postgres_create_directories" 
+
+    # ensure the postgres data directory exists and is owned by snap_daemon
+    logger "edgexfoundry:postgres: creating $POSTGRES_DIR"
+    mkdir -p "$POSTGRES_DIR"
+    chown -R snap_daemon:snap_daemon "$POSTGRES_DIR"
+    
+    # ensure the sockets dir exists and is owned by snap_daemon
+    logger "edgexfoundry:postgres: creating $SNAP_COMMON/sockets"
+    mkdir -p "$SNAP_COMMON/sockets"
+    chown -R snap_daemon:snap_daemon "$SNAP_COMMON/sockets"
+
+    # create the config directory used for the kong password
+    logger "edgexfoundry:postgres: creating $CONFIG_POSTGRES_PASSWORD_DIR"
+    mkdir -p $CONFIG_POSTGRES_PASSWORD_DIR 
+
+    # create directory for configuration
+    mkdir -p $CONFIG_POSTGRES_12_CONFIGFILE_DIR
+
+}
+
+postgres_create_postgres_database() {
+    logger "edgexfoundry:postgres: postgres_create_postgres_database" 
+    logger "edgexfoundry:postgres: initdb -D $POSTGRES_12_DB_DIR"  
+    "$SNAP/bin/drop-snap-daemon.sh" "$SNAP/usr/lib/postgresql/12/bin/initdb" -D "$POSTGRES_12_DB_DIR"
+}
+
+
+postgres_create_configuration_file() {
+
+    logger "edgexfoundry:postgres: postgres_create_configuration"
+    logger "edgexfoundry:postgres: setting up new configuration file at $CONFIG_POSTGRES_12_CONFIGFILE_DIR/$CONFIG_POSTGRES_CONFIGFILENAME"
+    
+    cp "$SNAP/config/postgres/$CONFIG_POSTGRES_CONFIGFILENAME" "$CONFIG_POSTGRES_12_CONFIGFILE_DIR/"
+
+    # do replacement of the $SNAP, $SNAP_DATA, $SNAP_COMMON environment variables in the config files
+    sed -i -e "s@\$SNAP_COMMON@$SNAP_COMMON@g" \
+        -e "s@\$SNAP_DATA@$CURRENT@g" \
+        -e "s@\$SNAP@$SNAP_CURRENT@g" \
+        "$CONFIG_POSTGRES_12_CONFIGFILE_DIR/$CONFIG_POSTGRES_CONFIGFILENAME" 
+    
+}
+
+postgres_remove_old_postgres() {
+    logger "edgexfoundry:postgres: postgres_remove_old_postgres"
+    rm -rf "$CONFIG_POSTGRES_10_CONFIGFILE_DIR"
+    "$SNAP/bin/drop-snap-daemon.sh" rm -rf "$POSTGRES_10_DIR"
+}
+
+# called from the security-proxy-post-setup script.
+# kong needs to be running for this to succeed, so we can't run this in the 
+# post-refresh hook
+kong_restore_backup() {
+    if [ -f "$SNAP_COMMON/kong.yaml" ]; then
+      logger "edgexfoundry:postgres: refresh - loading saved kong data"
+      cd $SNAP_COMMON
+      $SNAP/bin/deck diff > "diff.txt"
+      $SNAP/bin/deck sync | logger
+    fi
+}
+
+postgres_install() {
+     logger "edgexfoundry:postgres: postgres_install"
+ 
```

Comment:
Please add comment to mention that this is called from the install hook as well as the post refresh function.